test/integration/apimachinery/watch_restart_test.go

```diff
@@ -204,71 +204,74 @@ func TestWatchRestartsIfTimeoutNotReached(t *testing.T) {
 		},
 	}
 
-	for _, tmptc := range tt {
-		tc := tmptc // we need to copy it for parallel runs
-		t.Run(tc.name, func(t *testing.T) {
-			c, err := kubernetes.NewForConfig(config)
-			if err != nil {
-				t.Fatalf("Failed to create clientset: %v", err)
-			}
-
-			secret, err := c.CoreV1().Secrets(tc.secret.Namespace).Create(context.TODO(), tc.secret, metav1.CreateOptions{})
-			if err != nil {
-				t.Fatalf("Failed to create testing secret %s/%s: %v", tc.secret.Namespace, tc.secret.Name, err)
-			}
-
-			watcher, err, doneFn := tc.getWatcher(c, secret)
-			if err != nil {
-				t.Fatalf("Failed to create watcher: %v", err)
-			}
-			defer doneFn()
-
-			var referenceOutput []string
-			var output []string
-			stopChan := make(chan struct{})
-			stoppedChan := make(chan struct{})
-			go generateEvents(t, c, secret, &referenceOutput, stopChan, stoppedChan)
-
-			// Record current time to be able to asses if the timeout has been reached
-			startTime := time.Now()
-			ctx, cancel := watchtools.ContextWithOptionalTimeout(context.Background(), timeout)
-			defer cancel()
-			_, err = watchtools.UntilWithoutRetry(ctx, watcher, func(event watch.Event) (bool, error) {
-				s, ok := event.Object.(*corev1.Secret)
-				if !ok {
-					t.Fatalf("Received an object that is not a Secret: %#v", event.Object)
+	t.Run("group", func(t *testing.T) {
```

Comment:
What's the purpose of this outer `t.Run`? It doesn't change the control of parallelism.

Reply:
it ensures the outer setup/teardown code doesn't run until all the subtests are complete:

See https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks

> Run does not return until parallel subtests have completed, providing a way to clean up after a group of parallel tests:
> 
> ```
> func TestTeardownParallel(t *testing.T) {
>     // This Run will not return until the parallel tests finish.
>     t.Run("group", func(t *testing.T) {
>         t.Run("Test1", parallelTest1)
>         t.Run("Test2", parallelTest2)
>         t.Run("Test3", parallelTest3)
>     })
>     // <tear-down code>
> }
> ```

Reply:
Got it. I found https://github.com/golang/go/issues/17791 explaining the ordering better than the doc.